### DISCUSSION TITLE
how to run the wf-install script to install wayfire automatically

### DISCUSSION THREAD

**Discussion by Marietto2008** [WayfireWM/wayfire#2182] [General]
@ammen99 ; @soreau :

ok. Let's continue here :

```
[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> git config --global --add safe.directory /usr/home/marietto/Scaricati/wf-install/wayfire

error: can't block configuration file /home/marietto/.config/git/config: Permission denied

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> sudo git config --global --add safe.directory /usr/home/marietto/Scaricati/wf-install/wayfire
Password:

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> ./install.sh --prefix /opt/wayfire
Building Wayfire master
Installation prefix: /opt/wayfire

fatal: detected dubious ownership in repository at '/usr/home/marietto/Scaricati/wf-install/wayfire'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/home/marietto/Scaricati/wf-install/wayfire

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> git config --global --add safe.directory /usr/home/marietto/Scaricati/wf-install/wayfire

error: can't block configuration file /home/marietto/.config/git/config: Permission denied

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> sudo git config --global --add safe.directory /usr/home/marietto/Scaricati/wf-install/wayfire

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> ./install.sh --prefix /opt/wayfire
Building Wayfire master
Installation prefix: /opt/wayfire
fatal: detected dubious ownership in repository at '/usr/home/marietto/Scaricati/wf-install/wayfire'

To add an exception for this directory, call:

	git config --global --add safe.directory /usr/home/marietto/Scaricati/wf-install/wayfire

```

**Comment by ammen99** [comment]
I would suggest deleting the wf-install folder together with everything and re-cloning it as a normal user + run install.sh as a normal user again.

Also, your git permissions are messed up. You should fix that first - I think google should be able to help there ;)

**Comment by Marietto2008** [comment]
```
[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install] ==> # git clone https://github.com/WayfireWM/wf-install && cd wf-install

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install] ==>./install.sh --prefix /opt/wayfire

[62/62] Linking target src/panel/wf-panel
ninja: Entering directory `build'
[0/1] Installing files.
Installing subprojects/wayland-logout/wayland-logout to /opt/wayfire/bin
Installing src/panel/wf-panel to /opt/wayfire/bin
Installing src/background/wf-background to /opt/wayfire/bin
Installing src/dock/wf-dock to /opt/wayfire/bin
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/subprojects/wayland-logout/wayland-logout.1 to /opt/wayfire/share/man/man1
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/build/metadata/background.xml to /opt/wayfire/share/wayfire/metadata/wf-shell
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/metadata/dock.xml to /opt/wayfire/share/wayfire/metadata/wf-shell
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/metadata/panel.xml to /opt/wayfire/share/wayfire/metadata/wf-shell
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/data/wayfire.png to /opt/wayfire/share/wayfire/icons
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/data/wallpaper.jpg to /opt/wayfire/share/wayfire
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/build/meson-private/wf-shell.pc to /opt/wayfire/libdata/pkgconfig
Do you want to override the existing config file /home/marietto/.config/wayfire.ini [y/n]? y
cp: --backup=t is not a directory
```

**Reply by ammen99** [reply]
Can you try editing `install.sh` on line 149, move the `--backup=t` parameter to the beginning of the command (i.e immediately after `cp`), maybe there is a slight difference between the Linux and BSD versions of `cp`.

**Reply by Marietto2008** [reply]
Do you want to override the existing config file /home/marietto/.config/wayfire.ini [y/n]? y
cp: illegal option -- -
usage: cp [-R [-H | -L | -P]] [-f | -i | -n] [-alpsvx] source_file target_file
       cp [-R [-H | -L | -P]] [-f | -i | -n] [-alpsvx] source_file ... target_directory

This is how I have modified it :

cp --backup=t "${CONFIG_FILE}" "${DEFAULT_CONFIG_PATH}"

**Reply by Marietto2008** [reply]
Is the parameter --backup=t really important ? because I tried to remove it and I've got different errors later...(the same errors that I've got running the script as root) ; so the problem is not to run the script as root.

**Reply by Marietto2008** [reply]
I think this error is more serious than the previous one :

Do you want to override the existing config file /home/marietto/.config/wayfire.ini [y/n]? y
Do you want to override the existing config file /home/marietto/.config/wf-shell.ini [y/n]? y

**sed: 1: "/usr/home/marietto/Scar ...": extra characters at the end of h command**

**Reply by ammen99** [reply]
Oh, I see. `cp` does not have such an option at all on BSD, just checked the documentation as well. In this case @Marietto2008 please take a look at the install.sh file and modify/remove all Linux-specific stuff. I think you are the first one to try and report wf-install on *BSD which is why I did not recognize the bug at first.

About this particular instance, I think it is safe to just remove the `--backup=t` parameter, or even the whole command altogether, since you already have a wayfire.ini file. It is only meant for first-time installations.

The errors you were getting later were related to the `sed` command. I suppose we're also using something linux-specific there: if you can figure out how the commands can be changed (I have no idea about BSD, unfortunately) I would be happy to add the changes upstream as well.

**Reply by Marietto2008** [reply]
that's not an error caused by the fact that I use zsh. Infact I've changed the default shell from zsh to bash :

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]$ zsh -version
GNU bash, versione 5.2.21(0)-release (amd64-portbld-freebsd14.0)
Copyright (C) 2022 Free Software Foundation, Inc.
Licenza GPLv3+: GNU GPL versione 3 o successiva <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

but the error is still there :

sed: 1: "/usr/home/marietto/Scar ...": extra characters at the end of h command

**Reply by Marietto2008** [reply]
this is the counter proof :

```
[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> chsh -s bash
Password: 
chsh: user information updated

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]==> bash 

[marietto@marietto-I5 /usr/home/marietto/Scaricati/wf-install]$ bash -version
GNU bash, versione 5.2.21(0)-release (amd64-portbld-freebsd14.0)
Copyright (C) 2022 Free Software Foundation, Inc.
Licenza GPLv3+: GNU GPL versione 3 o successiva <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/data/wayfire.png to /opt/wayfire/share/wayfire/icons
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/data/wallpaper.jpg to /opt/wayfire/share/wayfire
Installing /usr/home/marietto/Scaricati/wf-install/wf-shell/build/meson-private/wf-shell.pc to /opt/wayfire/libdata/pkgconfig
Do you want to override the existing config file /home/marietto/.config/wayfire.ini [y/n]? n
Do you want to override the existing config file /home/marietto/.config/wf-shell.ini [y/n]? n

sed: 1: "/usr/home/marietto/Scar ...": extra characters at the end of h command

```

**Reply by ammen99** [reply]
I mean sed is not a function of the shell, sed is a utility program installed in /usr/bin and apparently on BSD, that utility program behaves differently than it does on Linux.

**Comment by soreau** [comment]
The problem is `drm-backend     : NO`, you will have to install the dependencies required to build wlroots with drm backend as wayfire requires it.

**Reply by Marietto2008** [reply]
I don't have the competences to convert that sed script for Linux in the corresponding script which works for BSD.

**Reply by ammen99** [reply]
@Marietto2008 I believe you meant to reply to my other comment but anyway the sed command replaces strings with the correct paths. You can comment out the lines and then replace the strings manually. For example this line:

```
    sed -i "s@^LD_.*@export LD_LIBRARY_PATH=${PREFIX}/${DEST_LIBDIR}:\$LD_LIBRARY_PATH@g" "$BUILDROOT/start_wayfire.sh"
```

Replaces the lines starting `LD_...` with the full version: `export LD_LIBRARY_PATH=$...`

**Reply by ammen99** [reply]
Also @Marietto2008 such things are easily google-able, one of the first answers explains how the command has to be adapted for *BSD: https://stackoverflow.com/questions/16745988/sed-command-with-i-option-in-place-editing-works-fine-on-ubuntu-but-not-mac

**Reply by Marietto2008** [reply]
so.  this :

sed -i "s@^LD_.*@export LD_LIBRARY_PATH=${PREFIX}/${DEST_LIBDIR}:\$LD_LIBRARY_PATH@g" "$BUILDROOT/start_wayfire.sh"

becomes this ?

export LD_LIBRARY_PATH="${PREFIX}/${DEST_LIBDIR}:\$LD_LIBRARY_PATH@g" "$BUILDROOT/start_wayfire.sh"

**Reply by ammen99** [reply]
@Marietto2008 try to change the script like this:

```
sed -i "s@...."
```

should become

```
sed -i '' "s@..."
```

in all places where we use sed. Take a look at the stackoverflow link for details.

**Reply by ammen99** [reply]
@Marietto2008 Ok :) I thought you had more knowledge in general since using *BSD is a rather niche thing, I'd expect most "beginners" to use Ubuntu/Fedora/Arch. In any case, what I wrote in the previous comment should be enough to fix that particular issue.

**Reply by Marietto2008** [reply]
I can have more knowledge,yes,I'm not stupid. But more you go deeply in the specific,less is the knowledge that I can obtain reading and trying to understand. You already know that the shell scripting is a very specific task. You should not be wondered by this.  Your fixes worked,except for the last two "sed" that I have converted from :

sed  -i "s@^Exec.*@Exec=$PREFIX/bin/startwayfire@g" "$BUILDROOT/wayfire.desktop"
sed  -i "s@^Icon.*@Icon=$PREFIX/share/wayfire/icons/wayfire.png@g" "$BUILDROOT/wayfire.desktop"

to :

sed '' -i "s@^Exec.*@Exec=$PREFIX/bin/startwayfire@g" "$BUILDROOT/wayfire.desktop"
sed '' -i "s@^Icon.*@Icon=$PREFIX/share/wayfire/icons/wayfire.png@g" "$BUILDROOT/wayfire.desktop"

but it didn't work. This is the error :

sed: -i: No such file or directory
sed: s@^Exec.*@Exec=/opt/wayfire/bin/startwayfire@g: No such file or directory

**Reply by Marietto2008** [reply]
oh sorry,my mistake. It should be :

sed -i '' "s@^Exec.@exec=$PREFIX/bin/startwayfire@g" "$BUILDROOT/wayfire.desktop"
sed -i '' "s@^Icon.@ICON=$PREFIX/share/wayfire/icons/wayfire.png@g" "$BUILDROOT/wayfire.desktop"

Installation done. Run /opt/wayfire/bin/startwayfire to start wayfire.

now the fun begins...

**Reply by Marietto2008** [reply]
```
#!/usr/local/bin/zsh
export LD_LIBRARY_PATH=/opt/wayfire/lib/$libdir
export PATH=/opt/wayfire/bin:$PATH
env WLR_NO_HARDWARE_CURSORS=1 LIBSEAT_BACKEND=consolekit2 ck-launch-session dbus-run-session /opt/wayfire/bin/./wayfire

```
=

```
failed to open /dev/dri/card0 : permission denied
failed to open /dev/dri/render0128 : permission denied
```

**Comment by soreau** [comment]
> ```
> failed to open /dev/dri/card0 : permission denied
> failed to open /dev/dri/render0128 : permission denied
> ```

You might need to add your user to the group associated with these files.

**Reply by Marietto2008** [reply]
I've added my username to video group :

`# pw groupmod video -m marietto`

but I've got the same error.

**Reply by ammen99** [reply]
@Marietto2008 Do you actually have a file called /dev/dri/render0128? On my machine, it is called /dev/dri/render**D**128

Also please post the full output of Wayfire.

**Reply by Marietto2008** [reply]
yes,sorry it was a typo : it is : /dev/dri/renderD128

log wayfire :

II 07-03-24 12:40:52.941 - [src/main.cpp:334] Starting wayfire version 0.9.0
II 07-03-24 12:40:52.944 - [libseat] [libseat/libseat.c:66] Seat opened with backend 'consolekit2'
EE 07-03-24 12:40:52.944 - [libseat] [libseat/backend/consolekit2.c:92] Ping failed: org.freedesktop.DBus.Error.AccessDenied: Rejected send message, 1 matched rules; type="method_call", sender=":1.377" (uid=1001 pid=60910 comm="") interface="org.freedesktop.DBus.Peer" member="Ping" error name="(unset)" requested_reply="0" destination="org.freedesktop.ConsoleKit" (uid=0 pid=2458 comm="")
II 07-03-24 12:40:52.944 - [backend/session/session.c:109] Successfully loaded libseat session
II 07-03-24 12:40:52.947 - [backend/backend.c:213] Found 1 GPUs
II 07-03-24 12:40:52.947 - [backend/drm/backend.c:202] Initializing DRM backend for /dev/dri/card0 (nvidia-drm)
II 07-03-24 12:40:52.947 - [backend/drm/drm.c:255] Found 4 DRM CRTCs
II 07-03-24 12:40:52.947 - [backend/drm/drm.c:213] Found 12 DRM planes
II 07-03-24 12:40:52.951 - [render/egl.c:206] Supported EGL client extensions: EGL_EXT_platform_base EGL_EXT_device_base EGL_EXT_device_enumeration EGL_EXT_device_query EGL_KHR_client_get_all_proc_addresses EGL_EXT_client_extensions EGL_KHR_debug EGL_KHR_platform_x11 EGL_EXT_platform_x11 EGL_EXT_platform_device EGL_MESA_platform_surfaceless EGL_EXT_explicit_device EGL_KHR_platform_wayland EGL_EXT_platform_wayland EGL_KHR_platform_gbm EGL_MESA_platform_gbm EGL_EXT_platform_xcb
EE 07-03-24 12:40:52.952 - [EGL] command: eglQueryDeviceStringEXT, error: EGL_BAD_PARAMETER (0x300c), message: "EGL_BAD_PARAMETER error: In internal function: Additional INFO may be available
"
II 07-03-24 12:40:52.952 - [EGL] command: eglQueryDeviceStringEXT, error: EGL_BAD_PARAMETER (0x300c), message: "EGL_BAD_PARAMETER error: In function eglQueryDeviceStringEXT(), string query failed at backend
"
EE 07-03-24 12:40:52.952 - [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to load driver"
EE 07-03-24 12:40:52.952 - [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to load driver"
EE 07-03-24 12:40:52.952 - [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to load driver"
EE 07-03-24 12:40:52.952 - [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "eglInitialize"
EE 07-03-24 12:40:52.952 - [render/egl.c:269] Failed to initialize EGL
EE 07-03-24 12:40:52.952 - [render/egl.c:572] Failed to initialize EGL context
EE 07-03-24 12:40:52.952 - [render/gles2/renderer.c:804] Could not initialize EGL
EE 07-03-24 12:40:52.952 - [src/main.cpp:131] Fatal error: Fatal error(SIGABRT)
EE 07-03-24 12:40:52.954 - #1  0x36b355 <_ZL14signal_handleri+0xb5> at /opt/wayfire/bin/./wayfire
EE 07-03-24 12:40:52.954 - #2  0x838eaa53f <pthread_sigmask+0x53f> at /lib/libthr.so.3
EE 07-03-24 12:40:52.954 - #3  0x838ea9afb <pthread_setschedparam+0x83b> at /lib/libthr.so.3
EE 07-03-24 12:40:52.954 - #4  0x8213de2d3 <__gxx_personality_v0+0x820fab923> at ???
EE 07-03-24 12:40:52.954 - #5  0x82c25e86a <thr_kill+0xa> at /lib/libc.so.7
EE 07-03-24 12:40:52.954 - #6  0x82c1d6f54 <_raise+0x34> at /lib/libc.so.7
EE 07-03-24 12:40:52.954 - #7  0x82c287fc9 <abort+0x49> at /lib/libc.so.7
EE 07-03-24 12:40:52.954 - #8  0x82c1ba2e1 <__assert+0x51> at /lib/libc.so.7
EE 07-03-24 12:40:52.954 - #9  0x36aedd <main+0xecd> at /opt/wayfire/bin/./wayfire
EE 07-03-24 12:40:52.954 - #10 0x82c1abafa <__libc_start1+0x12a> at /lib/libc.so.7

**Reply by Marietto2008** [reply]
my startup script :

#!/usr/local/bin/zsh
export LD_LIBRARY_PATH=/opt/wayfire/lib/$libdir
export PATH=/opt/wayfire/bin:$PATH
env WLR_NO_HARDWARE_CURSORS=1 LIBSEAT_BACKEND=consolekit2 ck-launch-session dbus-run-session opt/wayfire/bin/./wayfire (but also if I do only wayfire)

**Reply by Marietto2008** [reply]
I've found the solution ;

Modified /etc/devfs.rules as follows:

```
add path 'dri/*' mode 0666 group video
add path 'drm/*' mode 0666 group video
```

And add myself to group video.

**Reply by ammen99** [reply]
Oh, that's great :) Maybe the video group is set for these files automatically only under Linux and *BSD is again different here ... But in any case good that you found the cause.

**Reply by Marietto2008** [reply]
That's interesting to note that wayfire worked for some time without adding those paths on the devfs.rules and then,suddenly,it stopped working.

**Reply by Marietto2008** [reply]
Is there a way to place some icon / launchers in the middle of the screen,when you use wayfire,that when clicked,some kind of application is launched ? Every DM works like this. xfce,kde and so on. Wayfire has it ?

**Reply by soreau** [reply]
There are launchers you can configure for wf-panel but they are on the panel, not the middle of the screen. It's good you got the nvidia drivers working with wlroots on freebsd. Perhaps we should create a new discussion for general wayfire questions and configuration.
